### PR TITLE
Fixes #390: Simpler expression for INSTANCENAME

### LINKED ISSUE #390: Less error prone form of INSTANCENAME
The INSTANCENAME as WBEM URI is difficult to enter an error prone.  The use of  wildcard helps but is interactive.

One of our goals is to make pywbemcli as easily usable as possible, in particular make the command line input both intuitive and as free of specialized formats, characters, etc. as possible. Through experience with cimcli it became obvious that entering the CIMInstanceName parameter manually in a form acceptable to the WBEM URI standards was messy, partly because of the requirements for the double-quotes which must be escaped from most command line environments. The instance name is messy, involving multiple key properties and having to enter each with escaped double quotes for all non numerics caused users, especially casual users headaches in getting everything correct. This was compounded by the fact that a malformed instance name input on the command line simply generates an error that it was malformed.

There are several ways to make this easier including:

1. The wildcard which we already have which uses the classname with a wildcard "?* for the keybindings component and pywbemcli presents a list of possible known instance names.

2. Eliminate the need for the quotes as much as possible. We can make a form that is easier to input by eliminating the need for the quote marks in most cases.  This would use a parameter similar to the --property in instance create but it would be --key where each use would be a single key.  pywbemcli would assemble the CIMInstanceName from the individual keys including addin sayg quotes, etc.

3. Lead the user through the process step by step where pywbemcli would determine the key properties and present each key property with its type and ask the user to input a string representing the value

### PR DESCRIPTION
**Pull request by KSchopmeyer**
NOTE: This pr based on pr # 457, (fix tests as we move to pywbem 0.15.0)

Add option -k --key to allow keys of the instance name to be defined
separately and then aggragated by pywbemcli.

This allows us to handle the messy parts of creating CIMInstanceName
keybindings without having to involve the user in quote marks, etc.

Originally INSTANCENAME would have to be expressed as a WBEM_URL
complete with correctly placed double quotes.

CIM_Foo.InstanceId="blah"

This allows entering the INSTANCENAME as

CIM_Foo -k InstanceID=blah

### CODE REVIEW

File: docs/pywbemcli/features.rst [hunk 1]
```diff
@@ -361,3 +367,19 @@ interactive selection, as shown in the following example:
...
+.. _`Defining INSTANCENAME command argument with --key option`:
+
+`Defining INSTANCENAME command argument with --key option`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The INSTANCENAME may be specified by a combination of the namespace/classname
+as an argument with the --key option to define keybindings. Each --key option
+definition defines a single keybinding in the form name=value.   In general,
+the value component does not require the double quote that is required with
+WBEM_URL unless there are space characters in a string value.
```
Comment (andy-maier) [inline docs/pywbemcli/features.rst]: Change "WBEM_URL" to "WBEM URI"

File: pywbemtools/pywbemcli/_cmd_instance.py [hunk 2]
```diff
@@ -113,6 +115,14 @@
                       'supported. '
                       'Default: No properties modified.')]
 
+keybinding_key_option = [              # pylint: disable=invalid-name
+    click.option('-k', '--key', type=str, metavar='KEYNAME=VALUE',
+                 required=False, multiple=True,
+                 help='Value for a key in keybinding of CIMInstanceName. '
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_cmd_instance.py]: I suggest not to talk about pywbem CIM object classes (CIMInstanceName", but to say "CIM instance name" instead.

File: pywbemtools/pywbemcli/_cmd_instance.py [hunk 3]
```diff
@@ -113,6 +115,14 @@
...
+                      'May be specified multiple times. '
+                      'Allows defining keys without the issues of quotes. '
+                      'Default: No initial properties provided.')]
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_cmd_instance.py]: On "No initial properties provided": These are neither initial nor general properties. Suggest to change this to "No keybindings provided".

File: pywbemtools/pywbemcli/_cmd_instance.py [hunk 4]
```diff
@@ -939,7 +1041,7 @@ def cmd_instance_count(context, classname, options):
         except CIMError as ce:
             # allow processing to continue if no interop namespace
             if ce.status_code == CIM_ERR_NOT_FOUND:
-                click.echo('WARNING: {}'.format(ce))
+                warning_msg('{}'.format(ce))
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_cmd_instance.py]: This change is unrelated to the purpose of this PR. Please move to a separate PR.
Occurs multiple times in this file.

File: pywbemtools/pywbemcli/_common.py [hunk 5]
```diff
@@ -456,6 +456,51 @@ def create_cimproperty(cim_type, is_array, name, value_str):
     return CIMProperty(name, cim_value, cim_type)
 
 
+def create_ciminstancename(cim_class, kv_keys):
+    """
+    Create a cim instance from the input parameters.
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_common.py]: It creates a CIM instance name.

File: pywbemtools/pywbemcli/_common.py [hunk 6]
```diff
@@ -456,6 +456,51 @@ def create_cimproperty(cim_type, is_array, name, value_str):
...
+
+      Parameters:
+
+        cim_class_: (:Class:`pywbem.CIMClass`)
+            The class from which the CIMInstance is to be created
+
+        properties ():
+            A tuple of name/value pairs representing the properties and their
+            values that are to be constructed for the instance. Required
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_common.py]: Adjust the parameter description to the actual parameter.

File: pywbemtools/pywbemcli/_common.py [hunk 7]
```diff
@@ -456,6 +456,51 @@ def create_cimproperty(cim_type, is_array, name, value_str):
...
 
 
+def create_ciminstancename(cim_class, kv_keys):
+    """
+    Create a cim instance from the input parameters.
+
+      Parameters:
+
+        cim_class_: (:Class:`pywbem.CIMClass`)
+            The class from which the CIMInstance is to be created
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_common.py]: `CIMInstance` -> `CIMInstanceName`

File: pywbemtools/pywbemcli/_common.py [hunk 8]
```diff
@@ -456,6 +456,51 @@ def create_cimproperty(cim_type, is_array, name, value_str):
...
+            values that are to be constructed for the instance. Required
+
+      Returns: CIMInstanceName with namespace = None and host = None
+
+      Exceptions:
+        click.ClickException if Property name not found in class or if mismatch
+          of property type in class vs value component of kv pair
+    """
+    # TODO make usable if no class can be acquired because GetClass not
+    # implemented in server.
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_common.py]: This TODO is unrelated to the code in this function. The CIMCLass object is already passed in, so if there is an issue with GetClass not being implemented, it would be wherever that GetClass is executed, and that's where this TODO should be.

File: pywbemtools/pywbemcli/_common.py [hunk 9]
```diff
@@ -456,6 +456,51 @@ def create_cimproperty(cim_type, is_array, name, value_str):
...
+        name, value_str = parse_kv_pair(kv_property)
+        try:
+            cl_prop = cim_class.properties[name]
+        except KeyError:
+            raise click.ClickException('Property name "{}" not in class "{}".'
+                                       .format(name, cim_class.classname))
+
+        try:
+            if value_str.startswith('"') and value_str.endswith('"'):
+                value_str = value_str[1:-1]
```
Comment (andy-maier) [inline pywbemtools/pywbemcli/_common.py]: How about `value_str = value_str.strip('"')`?

File: tests/unit/test_common.py [hunk 10]
```diff
@@ -1,3 +1,4 @@
+# -*- coding: utf-8 -*-
```
Comment (andy-maier) [inline tests/unit/test_common.py]: The coding statement is only needed if the file contains plan Unicode characters. We should not do that however, but always escape them as `\uXXXX` or `\UXXXXXXXX`.

File: tests/unit/test_common.py [hunk 11]
```diff
@@ -1484,6 +1485,189 @@ def test_invalid_propname(self):
...
+                classname=u'CIM_Foo',
+                keybindings=NocaseDict(P1='Fred', P2='John', P3=123456)
+            ),
+        ),
+        None, None, True, ),
+    (
+        "Verify simple key creation with unicode char",
+        dict(
+            cls_kwargs=cls1,
+            kv_args=[u'P1=Fredäö'],
```
Comment (andy-maier) [inline tests/unit/test_common.py]: Please escape the Unicode characters.

File: tests/unit/test_common.py [hunk 12]
```diff
@@ -1484,6 +1485,189 @@ def test_invalid_propname(self):
...
+            namespace=None,
+            host=None,
+            exp_attrs=dict(
+                classname=u'CIM_Foo',
+                keybindings=NocaseDict(P1=u'Fredäö')
```
Comment (andy-maier) [inline tests/unit/test_common.py]: Please escape the Unicode characters.

File: docs/pywbemcli/features.rst (hunk 1 above)
Comment (KSchopmeyer) [reply docs/pywbemcli/features.rst]: DONE

File: pywbemtools/pywbemcli/_cmd_instance.py (hunk 2 above)
Comment (KSchopmeyer) [reply pywbemtools/pywbemcli/_cmd_instance.py]: DONE

File: pywbemtools/pywbemcli/_cmd_instance.py (hunk 3 above)
Comment (KSchopmeyer) [reply pywbemtools/pywbemcli/_cmd_instance.py]: DONE

File: pywbemtools/pywbemcli/_cmd_instance.py (hunk 4 above)
Comment (KSchopmeyer) [reply pywbemtools/pywbemcli/_cmd_instance.py]: DONE

File: pywbemtools/pywbemcli/_common.py (hunk 5 above)
Comment (KSchopmeyer) [reply pywbemtools/pywbemcli/_common.py]: DONE

File: pywbemtools/pywbemcli/_common.py (hunk 6 above)
Comment (KSchopmeyer) [reply pywbemtools/pywbemcli/_common.py]: DONE

File: pywbemtools/pywbemcli/_common.py (hunk 7 above)
Comment (KSchopmeyer) [reply pywbemtools/pywbemcli/_common.py]: DONE

File: pywbemtools/pywbemcli/_common.py (hunk 8 above)
Comment (KSchopmeyer) [reply pywbemtools/pywbemcli/_common.py]: Moved to the common funciton in _cmd_instance.py and extended the idea somewhat. Still a todo and marked future

**DONE**

File: pywbemtools/pywbemcli/_common.py (hunk 9 above)
Comment (KSchopmeyer) [reply pywbemtools/pywbemcli/_common.py]: 1. That would actually have to be more complex because it would have to account for escaped quotes inside the string.  This simply removes the lead and end if they both exist which is what we want to do because otherwise the create_cimvalue gets confused and adds more quotes.

2. No problem moving outside try block.  **DONE**

File: tests/unit/test_common.py (hunk 10 above)
Comment (KSchopmeyer) [reply tests/unit/test_common.py]: DONE

File: tests/unit/test_common.py (hunk 11 above)
Comment (KSchopmeyer) [reply tests/unit/test_common.py]: DONE

File: tests/unit/test_common.py (hunk 12 above)
Comment (KSchopmeyer) [reply tests/unit/test_common.py]: DONE

File: pywbemtools/pywbemcli/_common.py (hunk 9 above)
Comment (andy-maier) [reply pywbemtools/pywbemcli/_common.py]: `strip()` does remove only from start and end.
